File: ALGS200x/Week2/Challenge2-5/PY/fibmod.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
'''
Given `n m` on a line, compute fib(n) mod m.
'''
import sys

def compute_period(mod):
    '''Compute the Pisano period modulo `mod`'''
    seen = set()
    period = 0
    a = 0
    b = 1
    while (a, b) not in seen:
        period = period + 1
        seen.add((a, b))
        a, b = b, ((a + b) % mod)
    return period

def fib_mod(num, mod):
    '''naive computation for testing'''
    a, b = 0, 1
    count = 0
    while count < num:
        count = count + 1
        a, b = b, ((a + b) % mod)
    return a

def fib_mod_m(num, mod):
    '''compute fib(num) module mod'''
    period = compute_period(mod)
    num = num % period
    return fib_mod(num, mod)

def get_input():
    line = sys.stdin.readline()
    if not line:
        raise EOFError
    a, b = [int(x) for x in line.split()]
    return a, b

def run_tests():
    '''basic tests'''
    assert compute_period(2) == 3
    assert compute_period(3) == 8
    assert fib_mod_m(3, 3) == fib_mod(3, 3)
    assert fib_mod_m(5, 4) == fib_mod(5, 4)
    print("All tests pass.")

if __name__ == "__main__":
    #run_tests()
    while True:
        try:
            num, mod = get_input()
            print(fib_mod_m(num, mod))
        except EOFError:
            break

File: ALGS200x/Week2/Challenge2-5/CPP/fibmod.cpp
/*
 * Given 'n m' on a line, compute fib(n) mod m.
 */

#include <iostream>
#include <cassert>
#include <set>
#include <vector>

long compute_period(long mod)
{
  // To use unordered_set would be faster, but then it is necessary
  // to implement a hash function for <std::vector<long>>
  // If desired, that can be done: https://stackoverflow.com/a/29855973/1141805
  std::set<std::vector<long>> seen;
  long period = 0;
  long a = 0;
  long b = 1;
  long tmp;
  while (seen.count({a, b}) == 0)
  {
    period++;
    seen.insert({a, b});
    tmp = a;
    a = b;
    b = (a + tmp) % mod;
    //std::cout << a << " " << b << std::endl;
  }
  //std::cout << period << std::endl;
  return period;
}

long naive_fib_mod(long num, long mod)
{
  long a = 0;
  long b = 1;
  long count = 0;
  long tmp;
  while (count++ < num)
  {
    tmp = a;
    a = b;
    b = (tmp + a) % mod;
  }
  return a;
}

long fast_fib_mod(long num, long mod)
{
  long period = compute_period(mod);
  num = num % period;
  return naive_fib_mod(num, mod);
}

void run_tests()
{
  assert(compute_period(2) == 3);
  assert(compute_period(3) == 8);
  assert(fast_fib_mod(3, 3) == naive_fib_mod(3, 3));
  assert(fast_fib_mod(5, 4) == naive_fib_mod(5, 4));
  std::cout << "All tests pass." << std::endl;
}

int main()
{
  //run_tests();
  long a, b;
  while (std::cin >> a && std::cin >> b)
  {
    std::cout << fast_fib_mod(a, b) << std::endl;
  }
}
